### ISSUE TITLE
Allow for an alternate lookup path for sensitive information

### ISSUE THREAD

**Issue by pgehres** [ansible/ansible#4655]
In order to be able to share our playbooks with non-privileged users throughout our company we need to move all sensitive information (password, SSL certificates, etc) into a separate repository.  Unfortunately this isn't the easiest thing to do in the current architecture.

I think the easiest way to do this from a user perspective it to support an "alternative search path" that has the same structure as the playbook repo.  For example:

```
secret:
group_vars\
   nginx
host_vars\
   webserver.foo.com
roles\
   nginx\
      files\
         ssl.crt
         ssl.key

---
playbooks:
group_vars\
   nginx
host_vars\
   webserver.foo.com
roles\
   nginx\
      files\
         nginx.conf
      tasks
         main.yml
```

When running the main.yml for nginx, ansible would look in the "secret" repo first for a file and then search the normal code path. Variables from the "secret" repo would also take precedence so that you can provide a fully working role or playbook with default variables that are over-ridden by those in the "secret" repo.

**Comment by pgehres** [comment]
@mpdehaan @jimi-c @tgerla - as discussed yesterday

**Comment by mpdehaan** [comment]
Yep, I think this could be automatically added to all the file paths when we are looking for them.

I'd sugggest we probably add "extra_template_lookup_path" (default None) and extra_file_lookup_path
and this takes _lower_ priority than anything in the playbook/role/etc?

**Comment by pgehres** [comment]
My strong preference would be higher precedence to override things.

**Comment by mpdehaan** [comment]
BTW, you can somewhat do this now with the configurable role path on 1.4 devel

```
roles:
   - secret_bits
```

**Comment by pgehres** [comment]
That will work for variables, yes, but if I understand it correctly, I would need to apply that role to each host.  Additionally, I can't think of a way to make it work for files without putting the tasks that require those files in the secret_bits role as well, correct?

**Comment by pgehres** [comment]
https://github.com/pgehres/ansible/commit/dd2986c0795d871ccee724c5324b6c06ed1b3208 is a toy implementation of this.  @mpdehaan any thoughts on this direction? I realize that it complicates the logic a bit, but at least at this point I didn't to completely refactor the lookup code and this seemed like an easy way to bolt this feature on while preserving existing logic.

**Comment by mpdehaan** [comment]
We are going to be implementing the "vault" idea for encrypting data soon but this looks fine.

Apologies on late response as I sometimes miss github email notifications.

**Comment by pgehres** [comment]
Cool.  I will clean this code up and submit a pull request.  Thanks.

**Comment by mpdehaan** [comment]
Vault is now a part of 1.5.3 so I think we can consider this one solved, there is also the "roles_path" commit that allows loading things from a role called "my_secret_stuff" or so on, where variables could come in from that location, so the above should be possible via that as well, if for some reason usage of vault was not desired, or the vault content wanted to be in an additional repo as well.

**Comment by pgehres** [comment]
Just in case anyone stumbles across this, we ended up working around this.  Our goal was to be able to share roles throughout our organization to enable development and QA to use the same process that Ops uses in production. In the end, we created a "secrets" directory in the root of our playbooks (alongside roles) and store all of the sensitive files there. We store all passwords in group or host vars. (We are not using vault because change management and code review requires that they not be encrypted. Also, it makes it really hard to allow another team to use a different value.)
